content/blog/2022-05-17-object-oriented-features/index.md

```diff
@@ -0,0 +1,210 @@
+
+1. Method calls should be fast, since they are the core operation in the language
+2. The space overhead per object should be low. We are less concerned with space overhead on a per-class basis, since the number of objects is typically much larger than the number of classes.
+3. We would like to share inherited code across classes, since for a large class hierarchy, we might otherwise end up with many copies of the essentially the same code, leading to the dreaded “code bloat”.
+
+A straightforward approach that has been used in dynamically typed OO languages like Smalltalk is to simply search for the right method definition. Each object has a hidden field that points to its class object; a method call `o.m()` is implemented by walking up the class hierarchy, starting from o's class, and stopping once an implementation for method m is found. This implementation allows sharing inherited code, so it avoids code bloat, but it is much slower than a simple function call.
+
+For faster method dispatch, we need to precompute a data structure—a dispatch table. The key idea is to map different method names onto small integer indices, and to use those indices to look up the appropriate method code pointer in an array of code pointers. These arrays of code pointers go by various names. In C++, this array is called a vtbl, pronounced like “v-table”, standing for “virtual table". In other literature it is known as a selector table or dispatch vector.
+
+So consider the following classes:
+
+```
+class A           { f(); g(); }
+class B extends A { f(); h(); }
+```
+
+`A` and `B` should share the same call to `f`, `B` should inherit `g` from `A`, and `B` should also define `h`.
+
+We can put `f` and `g` in indicies 0 and 1 respectivley for both dispatch vectors and `h` is put in index 2 for the dispatch vector of `B`. Then, if there is a call to `f`, `g`, or `h`, we can get the pointer to the correct function in `O(1)` time by generating IR that resembles the following
+
+```
+MOV dv_ptr, o // Get the pointer to the dispatch vector since it is at index 0 of the object o
+MOV dv, [dv_ptr] // Get the dispatch vector
+CALL [dv + index] // Call a function based on its index in the dispatch vector
+```
+
+We build two classes to assist with building dispatch vectors:
+
+First we need to figure out when given some class `A`, and some function `f`, which classes `f` should we call. For example, if `f` is redefined by a subclass then we want to call the `f` of that subclass. But if `f` is defined by a super class we want to call `f` of that super class. This is all handled by the `DispatchVectorClassResolver`
+
+The next thing we need is given all methods for a given class, what index do they go in in the dispatch vector. For example if `A` defines `f` and subclass `B` also defines `f` they must be at the same index. the `DispatchVectorIndexResolver` handles this.
```

Comment:
As above. Is `DispatchVectorIndexResolver` just a global resource of the compiler, or is it a pass of some kind? What is it?